truffleruby/src/main/ruby/core/truffle/cext.rb

```diff
@@ -1111,6 +1111,52 @@ def rb_proc_new(function, value)
     end
   end
 
+  def foreign_call_with_block(function, *args, &block)
+        res = Truffle::Interop.execute(function, *args)
+  end
+
+  def rb_protect_with_block(function, arg, block)
+    res = nil;
+    pos = 0;
+    e = store_exception do
+      puts "Block is #{block}"
+      res = foreign_call_with_block(function, arg, &block)
+    end
+    unless e == nil
+      store = Thread.current[:__stored_exceptions__]
+      store = Thread.current[:__stored_exceptions__] = [] unless store
+      pso = store.index(nil) || store.size
+      store[pos] = e
+      pos += 1
+    end
+    [res, pos]
+  end
+
+  def rb_protect(function, arg)
+    res = nil;
+    pos = 0;
+    e = store_exception do
+      res = Truffle::Interop.execute(function, arg)
+    end
+    unless e == nil
+      store = Thread.current[:__stored_exceptions__]
+      store = Thread.current[:__stored_exceptions__] = [] unless store
+      pso = store.index(nil) || store.size
+      store[pos] = e
```

Comment:
I don't quite understand this logic, there are also typos.
Can we use a Hash or use it as a stack instead of fancy indexing?